### ISSUE TITLE
Bear idea: literal-objects sorter

### ISSUE THREAD

**Issue by NiklasMM** [coala/coala-bears#1419]
This idea just popped into my head:

I'd like my hard-coded objects/dicts to be sorted by key. Often I have something like:

```Python
default_menu_item = {
  "name": "some name",
  "calories": 1000,
  "price": 10,
  "name_addition": "main course",
  "price_currency": "euro"
}
```
These kinds of constructs exist in all kinds of languages and it would be nice to automatically sort them by key. The example above would look like this:

```Python
default_menu_item = {
  "calories": 1000,
  "name": "some name",
  "name_addition": "main course",
  "price": 10,
  "price_currency": "euro"
}
```
It would be nice, if we have a LiteralObjectSorterBear, which detects and fixes this. I don't know if there are tools that do this, but it could also be a very cool native bear. It would have to be designed such that it could deal with as many languages as possible.

**Comment by sils** [comment]
yes, ideally we'd probably build another bear on top of the annotationbear that can detect the dict syntax basically and then build our bear on that generic info or so. Then we can reuse the dict stuff for other bears later as well.

**Comment by Makman2** [comment]
(looks like coAST stuff...)

**Comment by jayvdb** [comment]
Other easy languages to support are json and yaml, but both are better 'ordered' by declaring a schema which defines a preferred key order. But this feature could be easily added to https://github.com/legoktm/jsonchecker and yamllint as a way of prototyping for coala to pick up on implementation complexities that need to be considered in our design.

Note that in Python and most other dict impl, re-ordering changes the dict which can introduce bugs, and even changes the data if there are duplicate keys. The last value for a key is used; others are discarded. pyflakes emits a warning for this as-of 1.4.

**Comment by aptrishu** [comment]
@sils @jayvdb 
I would rather do it with AST. We should not add this (dictionary range detection) functionality to AnnotationBear. AnnotationBear does too much, like it detects everything, no streamlined functionality.

**Comment by NiklasMM** [comment]
With AST, we would be limited to Python. I would rather have this be a language agnostic approach.

@jayvdb I agree, that we should use stable sorting, guaranteeing that double keys remain in the same order. Can you think of any other bugs this could introduce?

**Comment by aptrishu** [comment]
Actually No, I was talking about AST's in general.. We have an upcoming project (https://github.com/coala/coAST) . Integrating coala with ANTLR. So, we can do that for every language.

**Comment by NiklasMM** [comment]
Oh ok, I wasn't aware.

**Comment by aptrishu** [comment]
To be done after https://github.com/coala/coala-bears/issues/1431.

**Comment by sils** [comment]
I'm against using coAST for this. coAST won't be ready for ages and we have no idea if it ever will. Our Languages design is as modular as coala is and is sufficient for this usecase. I'm not saying we extend the AnnotationBear but we build on it: we build another bear that does annotations just like the AnnotationBear that depends on AnnotationBear and tells us about dicts in the code.

**Comment by aptrishu** [comment]
Agreed, I feel this..
So the idea right now is to build another bear which detects data structures such as dict, list/array..
It'll be dependent on AnnotationBear to take info of comments range to handle cases like

```Cython
# This is not a dict { 'key'='''
 print('hello world') 
# '''}
``` 

and then the sortingBear will be dependent on that Bear.

**Comment by Makman2** [comment]
the bear will do the same as coAST in principle...

why not making some kind of working prototype for Python in coAST? Maybe just use the built-in AST for now... just to get something there.

> I'm against using coAST for this. coAST won't be ready for ages and we have no idea if it ever will.

Damn let's drive it then :D

**Comment by IpshitaC** [comment]
@Udayan12167 can this be taken up for RGSoC?

**Comment by aptrishu** [comment]
As a project ?
Won't fit I think.

**Comment by IpshitaC** [comment]
cc @prachi1210

**Comment by jayvdb** [comment]
@Udayan12167 and I were talking about file sorting (i.e. the json/yaml implementation of this), as a type of problem that coala design doesnt have a good solution for. coala is very 'text diff' centric, which fails for binary files, can be a bit troublesome with varying EOLs, and also ends up with silly diffs for sorting of structured data, at least cases where most lines need to be reordered. In that case, chunks of removed and added lines are not helpful, as the lines only moved elsewhere. Showing diffs of those lines moving around is less than helpful.
`git` detects moved lines, but its ui is quite weak as it still is text diff centric.
It isnt a high priority problem, but we should be starting to think about alternative Result metadata that describes moved lines properly, and binary input files is going to require we have alt diff data structures.

**Comment by aptrishu** [comment]
>coala is very 'text diff' centric, which fails for binary files

@jayvdb can you give an example of such a case?
and yes its easy to normally detect moved blocks/lines.
and so we need to extend our Diff ? also there are cases of extra newlines messing around which creates a problem in detecting moved blocks.

**Comment by jayvdb** [comment]
Re binary files, we have a few binary bears with PRs, and a 'raw file' attribute on these bears.  This also allows support for non-Unicode text files.

A potential generic path forward for this is the nested languages project.

Rather than having a generic AST for everything, we can use Language-in-Language detection to find literal data structures within a source code, without knowing the syntax of the container file.

That is sort-of the opposite approach of `AnnotationBear`.

It would mean that bears using this approach might misunderstand file contents like

```
# This is not a dict { 'key'='''
 print('hello world') 
# '''}
```

But ... it would sort the keys.  Big deal.  If someone put that crap in your code, and you had configure a bear to look at those files, dont be surprised when it does what you want.  coala Ignore syntax can be used to exclude it, or the setting `ignore = filename` can be used.

The benefit is that it would catch a dict in a string in a JS file.

```js
json_data = `
{
  'key': 'value',
  'key2': 3
}`
```

An AST based approach would miss that, unless it also trys to parse the contents of any string it encounters.  But which parser should it use?
